shared/vendored/snakemake/versioning.smk

```diff
@@ -0,0 +1,163 @@
+"""
+Shared functions to be used within a Snakemake workflow for enforcing
+versions of dependencies the repo defines within its `nextstrain-pathogen.yaml`
+"""
+
+from os import path
+from sys import stderr
+from packaging.specifiers import SpecifierSet, InvalidSpecifier # snakemake dependency
+from packaging.version    import Version, InvalidVersion        # snakemake dependency
+from importlib.metadata import version as importlib_version, PackageNotFoundError
+from snakemake.common import __version__ as snakemake_version
+import subprocess
+import re
+
+class DependencyChecker():
+    def __init__(self, registration):
+        super().__init__()
+        self.declared_dependencies = registration.get('dependencies', {})
+        self.error_attrs = ["version_incompatibilities", "not_found_dependencies", "declaration_errors", "unexpected_errors"]
+        for attr in self.error_attrs:
+            setattr(self, attr, [])
+
+    def check(self):
+        for package_name, declared_spec in self.declared_dependencies.get('py', {}).items():
+            self.check_python_package(package_name, declared_spec)
+        for program_name, declared_spec in self.declared_dependencies.get('cli', {}).items():
+            self.check_cli_version(program_name, declared_spec)
+
+    def report_errors(self) -> bool:
+        if sum([len(getattr(self, attr)) for attr in self.error_attrs])==0:
+            print("All dependencies declared by this pathogen satisfied", file=stderr)
+            return False
+        
+        print(f"This pathogen declares dependencies which were not met.", file=stderr)
+        for attr in self.error_attrs:
+            errors = getattr(self, attr)
+            if len(errors)==0:
+                continue
+            print(attr.replace('_', ' ').upper() + ":")
+            for msg in errors:
+                print(f"\t{msg}", file=stderr)
+        return True
+
+    def check_python_package(self, name: str, declared_spec: str):
+        """
+        Check whether the installed python library *name* meets the specifier *spec_str*.
+        This uses importlib.metadata to check the available version which avoids importing
+        the top-level import. Any errors or if the declared dependency isn't satisfied are
+        logged.
+        """
+        try:
+            if name=='snakemake':
+                # importlib reports 0.0.0, so follow the approach of Snakemake's own min_version function
+                version = Version(snakemake_version)
```

Comment:
Huh, I can't replicate the 0.0.0 version...

```
$ nextstrain shell .
$ python 
Python 3.10.18 (main, Jul  1 2025, 10:55:01) [GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from importlib.metadata import version as importlib_version
>>> from packaging.version import Version
>>> Version(importlib_version("snakemake"))
<Version('7.32.3')>
```

Reply:
From my conda environment:

```console
$ snakemake --version
9.7.1
$ python
Python 3.12.2 | packaged by conda-forge | (main, Feb 16 2024, 20:54:21) [Clang 16.0.6 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from importlib.metadata import version as importlib_version
>>> from packaging.version import Version
>>> Version(importlib_version("snakemake"))
<Version('0.0.0')>
```

Reply:
> Then snakemake should work as a cli dep

`snakemake` was previously specified as a CLI dep!

> do we need to separate py vs cli

If we want the simplicity of not differentiating between these then we have to special-case the python version checking of snakemake or special-case a returned version of '0.0.0'. I've done the former.